zapcore/deferred_with.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2023 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zapcore
+
+import "sync"
+
+type deferredWithCore struct {
+	Core
+	sync.Once
+	fields []Field
+}
+
+var (
+	_ Core           = (*deferredWithCore)(nil)
+	_ leveledEnabler = (*deferredWithCore)(nil)
+)
+
+// NewDeferredWith wraps a Core with a "lazy" core that will.
+// only execute With operations if the logger is actually used
+func NewDeferredWith(core Core, fields []Field) Core {
+	return &deferredWithCore{
+		Core:   core,
+		fields: fields,
+	}
+}
+
+func (d *deferredWithCore) initOnce() {
+	d.Once.Do(func() {
+		d.Core = d.Core.With(d.fields)
+	})
+}
+
+func (d *deferredWithCore) With(fields []Field) Core {
```

Comment:
from internal discussion

I still wonder if this couldn't do something like
```
func (d *deferredWithCore) With(fields []Field) Core {
	return &deferredWithCore{
		Core:   core,
		fields: d.fields+fields, // paraphrasing, likely as a new copy
	}
```

it should be cheaper to both create, but also use later on?

Reply:
Yeah I like it, its even more performant on the Benchmark5WithsDeferredUsed, I just need to think through it a bit more carefully, is it correct in all cases?

```
goos: darwin
goarch: arm64
pkg: go.uber.org/zap
Benchmark5WithsUsed-10               	  496836	      2543 ns/op
Benchmark5WithsNotUsed-10            	  570384	      2251 ns/op
Benchmark5WithsDeferredUsed-10       	  770275	      1456 ns/op
Benchmark5WithsDeferredNotUsed-10    	 2621470	       464.9 ns/op
```

Reply:
Seems to be, added another test TestLoggerChainedWith

Reply:
I haven't given up.
I wish to write the tests that show the difference in behaviour first with the sync.Once core, and then port in the "mway" style pattern in zapfx context_core.go once I have confidence in the tests. Just been busy with day to day tasks :)

Reply:
I updated the tests - TestLoggerDeferredWithCapture

They are still using the "sync.Once" new core pattern, please check if the behaviour looks correct. Then I will try the mway pattern.

Reply:
Using your pattern, there is a subtle difference:

```
--- FAIL: TestLoggerDeferredWithCapture (0.00s)
    --- FAIL: TestLoggerDeferredWithCapture/deferred_logger_evaluates_at_time_of_logging,_and_stops_deferring_on_children (0.00s)
        logger_test.go:221:
            	Error Trace:	/Users/qjeremy/zap/logger_test.go:221
            	Error:      	Not equal:
            	            	expected: zap.res{A:[]int{2}, B:[]int{2}, C:[]int{2}, D:[]int{3}, E:[]int{4}}
            	            	actual  : zap.res{A:[]int{4}, B:[]int{4}, C:[]int{2}, D:[]int{4}, E:[]int{4}}

            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -2,6 +2,6 @@
            	            	  A: ([]int) (len=1) {
            	            	-  (int) 2
            	            	+  (int) 4
            	            	  },
            	            	  B: ([]int) (len=1) {
            	            	-  (int) 2
            	            	+  (int) 4
            	            	  },
            	            	@@ -11,3 +11,3 @@
            	            	  D: ([]int) (len=1) {
            	            	-  (int) 3
            	            	+  (int) 4
            	            	  },
            	Test:       	TestLoggerDeferredWithCapture/deferred_logger_evaluates_at_time_of_logging,_and_stops_deferring_on_children
```